---
title: "NAIR: Network Analysis of Immune Repertoire"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{NAIR: Network Analysis of Immune Repertoire}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center", 
  out.width = "100%",
  fig.width = 9, fig.height = 7
)
```


The `NAIR` package facilitates network analysis of the adaptive immune repertoire based on similarities among the receptor sequences. It implements methods from the following paper:

Hai Yang, Jason Cham, Zenghua Fan, Brian Neal, Tao He and Li Zhang. “NAIR: Network Analysis of Immune Repertoire” (Submitted).


### What can `NAIR` do?

`NAIR` allows the user to:

* Perform general network analysis on immune repertoire sequence (RepSeq) data, including computing local and global network properties of nodes and clusters
* Search across multiple RepSeq samples for:
    * Clones/clusters associated to a clinical outcome
    * Public clones/clusters 
* Generate customized visualizations of the immune repertoire network
* Perform further downstream analysis


### What data does `NAIR` support?

`NAIR` supports bulk and single-cell immune repertoire sequence data for T-cell or B-cell receptors (TCR or BCR).

* **Single-cell data:** Each row is a single T-cell/B-cell
* **Bulk data:** Each row is a distinct TCR/BCR clone (unique combination of V-D-J genes and nucleotide sequence) and typically includes a corresponding measurement of clonal abundance (e.g., clone count and clone frequency/fraction)


### How does `NAIR` model the immune repertoire as a network?

* Each TCR/BCR cell (single-cell data) or clone (bulk data) is modeled as a node (vertex) in the network
* For each node, we consider the corresponding receptor sequence (nucleotide or amino acid)
* For each pair of nodes, we measure the similarity in their receptor sequences (using the Hamming or Levenshtein distance)
* An edge is drawn between two nodes if the distance is below a specified threshold 
* **(Added in version 0.0.9018)** For single-cell data, sequences from two chains (e.g., alpha chain and beta chain) can be jointly used to determine similarity between cells, where cells are similar if and only if the sequences in both chains are similar.




# The `buildRepSeqNetwork()` function

General network analysis on RepSeq data is performed using the `buildRepSeqNetwork()` function. This function does the following:

* Builds the network graph for the immune repertoire
* Computes desired network properties
* Prints a customized `ggraph` plot of the network graph
* Returns meta-data for the TCR/BCR (nodes) in the network, including biological as well as network properties
* Also returns the network `igraph` and adjacency matrix, as well as the `ggraph` plot object, for use in downstream analysis

See [this vignette](buildRepSeqNetwork.html) for a tutorial on the function's usage.



# Network Visualization

The `buildRepSeqNetwork()` function and other top-level functions in the `NAIR` package include numerous optional arguments that allow customization of the network visualization. See [this vignette](network_visualization.html) for further details.



# Downstream Analysis

The output returned by `buildRepSeqNetwork()` and other top-level `NAIR` functions can be used to facilitate further downstream analysis with the help of auxiliary functions within the `NAIR` package. See [this vignette](downstream_analysis.html) for further details.



# Finding Associated Clusters

A set of functions has been provided to search across samples for TCR/BCR clones associated to a sample- or subject-level binary outcome/characteristic. An example usage is seen below. A detailed tutorial and documentation are forthcoming.

See [this vignette](associated_clusters.html) for a tutorial on the function's usage.



# Finding Public Clusters

A set of functions has been provided to search across samples for public clones. An example usage is seen below. A detailed tutorial and documentation are forthcoming.

See [this vignette](public_clusters.html) for a tutorial on the function's usage.
