---
title: "NAIR: Network Analysis of Immune Repertoire"
output:
  html_document:
    toc: yes
    toc_float: yes
    df_print: paged
---

The `NAIR` package facilitates network analysis of the adaptive immune repertoire based on similarities among the receptor sequences. It implements custom pipelines developed in the following paper:

Hai Yang, Jason Cham, Zenghua Fan, Brian Neal, Tao He and Li Zhang. "Network Analysis of Immune Repertoire (NAIR) with Advanced Machine Learning Techniques." In: Briefings in Bioinformatics (under review).


### What can it do?

`NAIR` allows the user to:

* Perform general network analysis on immune repertoire sequence (RepSeq) data, including computing local and global network properties of nodes and clusters
* Search across multiple RepSeq samples for:
* Clones/clusters associated to a clinical outcome
* Public clones/clusters 
* Perform further downstream analysis 


### What data does it support?

`NAIR` supports bulk and single-cell immune repertoire sequence data for T-cell or B-cell receptors (TCR or BCR).

* **Single-cell data:** Each row is a single TCR/BCR
* **Bulk data:** Each row is a distinct TCR/BCR clone (unique combination of V-D-J genes and nucleotide sequence) and typically includes a corresponding measurement of clonal abundance (e.g., count or frequency)


### How does it model the immune repertoire as a network?

* Each TCR/BCR cell (single-cell data) or clone (bulk data) is modeled as a node (vertex) in the network
* For each node, we consider the corresponding receptor sequence (nucleotide or amino acid)
* For each pair of nodes, we measure the similarity in their receptor sequences (using the Hamming or Levenshtein distance)
* An edge is drawn between two nodes if the distance is below a specified threshold 



# The `buildRepSeqNetwork()` function

General network analysis on RepSeq data is performed using the `buildRepSeqNetwork()` function. This function does the following:

* Builds the network graph for the immune repertoire
* Computes desired network properties
* Prints a customized `ggraph` plot of the network graph
* Returns meta-data for the TCR/BCR (nodes) in the network, including both biological and network properties
* If desired for downstream analysis, can also return the network `igraph` and adjacency matrix, as well as the `ggraph` plot object


For demonstration purposes, we load a public data set 

```{r }
if (!require("BiocFileCache", quietly = TRUE)) { 
  if (!require("BiocManager", quietly = TRUE)) { 
    install.packages("BiocManager") 
  }
  BiocManager::install("BiocFileCache") 
}
library(NAIR)
library(BiocFileCache)
bfc <- BiocFileCache(ask = FALSE)
data_file <- bfcrpath(bfc, file.path(
  "http://cf.10xgenomics.com/samples/cell-vdj/3.1.0",
  "vdj_v1_hs_pbmc3/vdj_v1_hs_pbmc3_t_filtered_contig_annotations.csv"))
data <- read.csv(data_file, stringsAsFactors = FALSE)


# load data sets
# dir_main <-
#   dirname(dirname(rstudioapi::getActiveDocumentContext()$path))
# dir_data <- file.path(dir_main, "input_data")
# dir_out <- file.path(dir_data, "dummy_output")
# 
# data <- read.table(
#   file.path(dir_data, "TRB-Pt-10-2-500ng-15-04-2020-gDNA_S48.clones.txt"),
#   sep = '\t', header = TRUE)

```


The RepSeq data has the following dimensions and variables:
```{r }
dim(data)
(names(data))
head(data)
```



# Basic Usage

We first demonstrate the `buildRepSeqNetwork()` function using the default settings. 

The first argument, `data`, specifies the data frame containing the rep-seq data. Both bulk and single-cell rep-seq data are applicable. 

Each row of `data` is assumed to correspond to a single TCR/BCR. The nucleotide or amino acid sequence for each TCR/BCR (row) need not be unique.

The following arguments each specify the column name or number of `data` corresponding to a particular variable:

* `nucleo_col` contains the TCR/BCR nucleotide sequence
* `amino_col` contains the amino acid sequence corresponding to the TCR/BCR nucleotide sequence
* `count_col` contains the clone count (abundance) of the TCR/BCR in the current row (for bulk rep-seq data)
* `freq_col` contains the clone frequency/fraction (for bulk rep-seq data)
* `vgene_col` contains the V gene corresponding to the TCR/BCR in the current row
* `dgene_col` contains the D gene corresponding to the TCR/BCR in the current row
* `jgene_col` contains the J gene corresponding to the TCR/BCR in the current row
* `cdr3length_col` contains the length of the TCR/BCR CDR3 nucleotide sequence


```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    vgene_col = "bestVGene", 
    dgene_col = "bestDGene",
    jgene_col = "bestJGene", 
    cdr3length_col = "lengthOfCDR3")
```

By default:

* The network is based on similarity in amino acid sequence, as measured by the Hamming distance.
* Prior to building the network, the input data is filtered to remove rows whose TCR/BCR sequence contains fewer than 3 characters.
* When building the network, nodes with zero network degree (those that do not share an edge/adjacency with any other nodes in the network) are dropped.

A plot of the network graph is automatically printed. 

The function returns a copy of `data` containing only those columns specified in the function arguments, and only those rows corresponding to nodes that remain in the final network.

```{r}
dim(network)
head(network)
```

In the following sections we will demonstrate how the optional arguments of the `buildRepSeqNetwork()` function can be used to customize its behavior and output.


# Data/Clone Sequence Arguments

## keep additional data columns

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    vgene_col = "bestVGene", 
    dgene_col = "bestDGene",
    jgene_col = "bestJGene", 
    cdr3length_col = "lengthOfCDR3",
    other_cols = c("allVHits", "allDHits", "allJHits"))
```


```{r}
dim(network)
head(network)
```


## aggregate clones by unique amino acid sequence

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE)
```

Legend reflects aggregate clone count.

```{r}
dim(network)
head(network[order(network$AggregatedCloneCount, decreasing = TRUE) , ])
```

Fewer columns are kept when clones are aggregated; a variable for unique clone count is added, which counts the number of rows that were aggregated for each clone sequence.


## Removing seqs with special characters

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    drop_chars = "[*|_]",
    aggregate_identical_clones = TRUE)
```



## Using nucleotide sequence

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    vgene_col = "bestVGene", 
    dgene_col = "bestDGene",
    jgene_col = "bestJGene", 
    cdr3length_col = "lengthOfCDR3",
    clone_seq_type = "nucleotide"
  )
```

Default plot title reflects the change.

```{r}
dim(network)
head(network)
```


# Network Arguments

## Using Levenshtein distance

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    dist_type = "levenshtein"
  )
```

Default plot subtitle reflects the change.


```{r}
dim(network)
head(network[order(network$AggregatedCloneCount, decreasing = TRUE) , ])
```



## Using Euclidean distance on Atchley factor embedded values

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    drop_chars = "[*|_]",
    aggregate_identical_clones = TRUE,
    dist_type = "euclidean_on_atchley"
  )
```



## Add (default) network stats

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    node_stats = TRUE)
```

Default node coloring is now by degree since it is available.

```{r}
dim(network)
head(network[order(network$AggregatedCloneCount, decreasing = TRUE) , ])
```

Node stats are added to the data.

## Add all possible network stats

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    node_stats = TRUE,
    stats_to_include = "all")
```

```{r}
dim(network)
head(network[order(network$AggregatedCloneCount, decreasing = TRUE) , ])
```



## Add custom network stats

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    node_stats = TRUE,
    stats_to_include = node_stat_settings(cluster_id = TRUE,
                                          transitivity = FALSE,
                                          closeness = TRUE,
                                          centrality_by_closeness = TRUE,
                                          centrality_by_eigen = FALSE))
```

```{r}
dim(network)
head(network[order(network$AggregatedCloneCount, decreasing = TRUE) , ])
```






## Add cluster stats

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    cluster_stats = TRUE)
```

```{r}
typeof(network)
names(network)
dim(network$node_data)
dim(network$cluster_stats)
```

```{r}
head(network$node_data[order(network$node_data$AggregatedCloneCount, 
                             decreasing = TRUE) , ])
```

```{r}
head(network$cluster_stats)
```

# Plot Settings

## Custom Plot Title/Subtitle

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    plot_title = "Sample: TRB-Pt-10-2\nNetwork on CDR3 amino acid sequence")
```



## Custom variables for node size/color

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    cluster_stats = TRUE,
    size_nodes_by = "cloneFraction",
    color_nodes_by = "cluster_id")
```

Notice that "cluster_id" or other node-level stats to be added to the data can be used to color or size the nodes; users are not restricted only to columns initially present in the data.

## Custom color scheme

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    node_stats = TRUE,
    color_nodes_by = "eigen_centrality",
    color_scheme = "plasma")
```




## Multiple color variables for nodes

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    node_stats = TRUE,
    color_nodes_by = c("centrality_by_eigen", "centrality_by_betweenness"),
    color_scheme = "plasma")
```

One plot is generated for each color variable.

A different color scheme can be specified for each variable if desired:

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    node_stats = TRUE,
    stats_to_include = "all",
    color_nodes_by = c("cluster_id", "degree"),
    color_scheme = c("turbo", "plasma"))
```



## Adjust node size range

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    node_stats = TRUE,
    stats_to_include = "all",
    size_nodes_by = "eigen_centrality",
    node_size_limits = c(0.1, 5),
    color_nodes_by = "centrality_by_closeness")
```


## Fixed node sizes

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    size_nodes_by = 3
  )
```



## Adjust edge width

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    edge_width = 1 # default is 0.3
  )
```


## Custom legend text

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    size_title = "aggregate clone count",
    color_title = "aggregate clone count",)
```



With multiple plots:

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_identical_clones = TRUE,
    node_stats = TRUE,
    stats_to_include = "all",
    color_nodes_by = c("cluster_id", "degree"),
    color_scheme = c("turbo", "plasma"),
    size_title = "aggregate clone count",
    color_title = c("cluster ID", "network degree"))
```




# Save/Customize Output

## Save data & plot

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    vgene_col = "bestVGene", 
    dgene_col = "bestDGene",
    jgene_col = "bestJGene", 
    cdr3length_col = "lengthOfCDR3",
    node_stats = TRUE,
    color_nodes_by = c("cloneCount", "degree"),
    output_dir = dir_out)
```

If cluster statistics are computed, they are automatically saved as well when an output directory is provided:

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    vgene_col = "bestVGene", 
    dgene_col = "bestDGene",
    jgene_col = "bestJGene", 
    cdr3length_col = "lengthOfCDR3",
    cluster_stats = TRUE,
    output_dir = dir_out)
```

## Save everything

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    vgene_col = "bestVGene", 
    dgene_col = "bestDGene",
    jgene_col = "bestJGene", 
    cdr3length_col = "lengthOfCDR3",
    output_dir = dir_out,
    save_all = TRUE)
```
Using this option also writes files for the igraph and the adjacency matrix.



## Custom filenames

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    vgene_col = "bestVGene", 
    dgene_col = "bestDGene",
    jgene_col = "bestJGene", 
    cdr3length_col = "lengthOfCDR3",
    output_dir = dir_out,
    save_all = TRUE,
    data_outfile = "custom_data_file.csv",
    plot_outfile = "custom_plot_file.pdf",
    cluster_outfile = "custom_cluster_file.csv",
    igraph_outfile = "custom_igraph_file.txt", 
    matrix_outfile = "custom_matrix_file.mtx")
```


## Adjust pdf dimensions



```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    vgene_col = "bestVGene", 
    dgene_col = "bestDGene",
    jgene_col = "bestJGene", 
    cdr3length_col = "lengthOfCDR3",
    output_dir = dir_out,
    plot_width = 14,
    plot_height = 11)
```


## Return everything in a list


```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    vgene_col = "bestVGene", 
    dgene_col = "bestDGene",
    jgene_col = "bestJGene", 
    cdr3length_col = "lengthOfCDR3",
    cluster_stats = TRUE,
    return_all = TRUE)
```

```{r}
typeof(network)
names(network)
```

