---
title: "Build RepSeq Network (Demo Argument Usage and Output)"
output:
  html_document:
    toc: yes
    toc_float: yes
    df_print: paged
---

## Load data
```{r }
library(RepSeqNetworkAnalysis)

# load data sets
dir_main <- dirname(dirname(rstudioapi::getActiveDocumentContext()$path))
dir_data <- file.path(dir_main, "data")
data <- read.table(
  file.path(dir_data, "TRB-Pt-10-2-500ng-15-04-2020-gDNA_S48.clones.txt"),
  sep = '\t', header = TRUE)

```


Check which columns correspond to nucleotide sequence, amino acid sequence, count and frequency:
```{r }
dim(data)
(names(data))
head(data)
```



## Default arguments

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    vgene_col = "bestVGene", 
    dgene_col = "bestDGene",
    jgene_col = "bestJGene", 
    cdr3length_col = "lengthOfCDR3")
```

Nodes are colored by clone count.

```{r}
dim(network)
head(network)
```


## With clones aggregated by unique amino acid sequence

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_reads = TRUE)
```

Legend reflects aggregate clone count.

```{r}
dim(network)
head(network[order(network$aggCloneCount, decreasing = TRUE) , ])
```

Fewer columns are kept when clones are aggregated; number of reads is added.

## Using nucleotide sequence

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    vgene_col = "bestVGene", 
    dgene_col = "bestDGene",
    jgene_col = "bestJGene", 
    cdr3length_col = "lengthOfCDR3",
    clone_seq_type = "nucleotide"
  )
```

Default plot title reflects the change.

```{r}
dim(network)
head(network)
```


## Using levenshtein distance

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_reads = TRUE,
    dist_type = "levenshtein"
  )
```

Default plot subtitle reflects the change.


```{r}
dim(network)
head(network[order(network$aggCloneCount, decreasing = TRUE) , ])
```




## Add (default) network stats

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_reads = TRUE,
    node_stats = TRUE)
```

Default node coloring is now by degree since it is available.

```{r}
dim(network)
head(network[order(network$aggCloneCount, decreasing = TRUE) , ])
```

Node stats are added to the data.

## Add all possible network stats

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_reads = TRUE,
    node_stats = TRUE,
    stats_to_include = "all")
```

```{r}
dim(network)
head(network[order(network$aggCloneCount, decreasing = TRUE) , ])
```



## Add custom network stats

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_reads = TRUE,
    node_stats = TRUE,
    stats_to_include = node_stat_settings(cluster_id = TRUE,
                                          transitivity = FALSE,
                                          closeness = TRUE,
                                          centrality_by_closeness = TRUE,
                                          centrality_by_eigen = FALSE))
```

```{r}
dim(network)
head(network[order(network$aggCloneCount, decreasing = TRUE) , ])
```






## Add cluster stats

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_reads = TRUE,
    cluster_stats = TRUE)
```

```{r}
typeof(network)
names(network)
dim(network$node_data)
dim(network$cluster_info)
```

```{r}
head(network$node_data[order(network$node_data$aggCloneCount, 
                             decreasing = TRUE) , ])
```

```{r}
head(network$cluster_info)
```



## Custom Plot Title/Subtitle

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_reads = TRUE,
    plot_title = "Sample: TRB-Pt-10-2\nNetwork on CDR3 amino acid sequence")
```



## Custom variables for node size/color

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_reads = TRUE,
    cluster_stats = TRUE,
    size_nodes_by = "cloneFraction",
    color_nodes_by = "cluster_id")
```

Notice that "cluster_id" or other node-level stats to be added to the data can be used to color or size the nodes; users are not restricted only to columns initially present in the data.


## Adjust node size range

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_reads = TRUE,
    node_stats = TRUE,
    stats_to_include = "all",
    size_nodes_by = "eigen_centrality",
    node_size_limits = c(0.1, 5),
    color_nodes_by = "centrality_by_closeness")
```


## Custom color scheme

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_reads = TRUE,
    node_stats = TRUE,
    color_nodes_by = "eigen_centrality",
    color_scheme = "plasma")
```




## Multiple color variables for nodes

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_reads = TRUE,
    node_stats = TRUE,
    color_nodes_by = c("centrality_by_eigen", "centrality_by_betweenness"),
    color_scheme = "plasma")
```

One plot is generated for each color variable.

A different color scheme can be specified for each variable if desired:

```{r fig.width = 10, fig.height = 10, out.width = "100%"}
# build network
network <- 
  buildRepSeqNetwork(
    data = data,
    nucleo_col = "nSeqCDR3",
    amino_col = "aaSeqCDR3",
    count_col = "cloneCount",
    freq_col = "cloneFraction",
    aggregate_reads = TRUE,
    node_stats = TRUE,
    stats_to_include = "all",
    color_nodes_by = c("cluster_id", "degree"),
    color_scheme = c("turbo", "plasma"))
```